contracts/compressor/IntentCompressor.sol

```diff
@@ -0,0 +1,336 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.10;
+
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import {IInbox} from "../interfaces/IInbox.sol";
+import {IIntentSource} from "../interfaces/IIntentSource.sol";
+import {Intent, Call, TokenAmount, Route, Reward} from "../types/Intent.sol";
+
+struct EncodedIntent {
+    uint8 sourceChainIndex;
+    uint8 destinationChainIndex;
+    // Reward token
+    uint8 rewardTokenIndex;
+    uint48 rewardAmount;
+    // Route token
+    uint8 routeTokenIndex;
+    uint48 routeAmount;
+    // Expiry duration
+    uint24 expiryDuration;
+    // Salt
+    bytes8 salt;
+}
+
+struct EncodedFulfillment {
+    uint8 sourceChainIndex;
+    uint8 destinationChainIndex;
+    // Reward token
+    uint8 routeTokenIndex;
+    uint48 routeAmount;
+    // Prove type (INSTANT = 0, BATCH = 1)
+    uint8 proveType;
+    // Claimant
+    address claimant;
+}
+
+contract IntentCompressor {
```

Comment:
I think it should be a library instead of contract, for gas efficiency we would build this directly into `Inbox`, `IntentSource`.

```
type PackedIntent is bytes32;

library IntentPacking {
    function decode(PackedIntent packed) pure returns (EncodedIntent memory) {
    
    }
    
    function publishAndFund(PackedIntent packed, address creator, address receiver) returns (bytes32 intentHash);
    
    // ...other intent methods?
}

Reply:
Totally, great idea! I build it this way!